docs: poc document for consistent batching configuration in log agent (#3402)
kyma-project/telemetry-manager: docs/contributor/pocs/consistent-batching-across-components/01-stanza-filelog-batching.md

# Investigate Stanza Batching Behavior and the SynchronousLogEmitter Feature Gate

## Provide Context

This investigation examines the `stanza.synchronousLogEmitter` feature gate, introduced in v0.122.0 at alpha stage, in the OpenTelemetry Collector's stanza package (used by the `filelogreceiver`). The gate prevents data loss on non-graceful shutdown by replacing the asynchronous `BatchingLogEmitter` with a synchronous passthrough emitter at the adapter level.

This investigation targets collector version 0.149.0, tailing Kubernetes CRI container logs in the `/var/log/pods/` directory structure.

Repository: `github.com/open-telemetry/opentelemetry-collector-contrib`
Package: `pkg/stanza`
Feature gate ID: `stanza.synchronousLogEmitter`
Reference issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35456

## Understand the Architecture

> The `filelog` receiver uses the stanza framework to tail log files. The data path is:

Reviewer: ```suggestion
The `filelog` receiver uses the stanza framework to tail log files. The data path is the following:
```

Low throughput: 1 log/s per Pod, 10 Pods, 300s (2,930 lines written). High throughput: 100 logs/s per Pod, 10 Pods, 2146s, queue saturated (147,956 lines written).

| Metric                             | Low: Sync | Low: Batching | High: Sync | High: Batching |
|------------------------------------|-----------|---------------|------------|----------------|
| Export requests (rpc_server count) | 2,930     | 297           | 2,389      | 2,259          |
| Records delivered                  | 2,930     | 2,930         | 147,256    | 146,656        |
| Avg records / filelog batch        | 1.0       | 9.8           | 61.6       | 65             |

At low throughput, readers rarely fill their 100-entry batch, so the `SynchronousLogEmitter` passes through many small batches while the `BatchingLogEmitter` coalesces them, producing a ~10x difference in export requests.

At high throughput, readers saturate at ~100 entries per batch regardless of emitter mode, collapsing the difference. The number of records delivered differs due to backpressure causing data loss.

### Finding 3: The Container Parser's Internal Emitter Is a Blind Spot for the Feature Gate

> The code checks the `stanza.synchronousLogEmitter` feature gate in only one place: `adapter/factory.go` line 79. The container parser hardcodes `helper.NewBatchingLogEmitter` in `operator/parser/container/config.go` line 84 without checking the gate.

Reviewer: ```suggestion
The code checks the `stanza.synchronousLogEmitter` feature gate only in `adapter/factory.go` line 79. The container parser hardcodes `helper.NewBatchingLogEmitter` in `operator/parser/container/config.go` line 84 without checking the gate.
```

# Investigate Stanza Batching Behavior and the SynchronousLogEmitter Feature Gate

## Provide Context

> This investigation examines the `stanza.synchronousLogEmitter` feature gate, introduced in v0.122.0 at alpha stage, in the OpenTelemetry Collector's `filelog` receiver. The gate prevents data loss on non-graceful shutdown by replacing the asynchronous `BatchingLogEmitter` with a synchronous passthrough emitter at the adapter level.

Reviewer 2: ```suggestion
This investigation examines the `stanza.synchronousLogEmitter` feature gate, introduced in v0.122.0 at alpha stage, in the OpenTelemetry Collector's `filelog` receiver. To prevent data loss on non-graceful shutdown, the gate replaces the asynchronous `BatchingLogEmitter` with a synchronous passthrough emitter at the adapter level.
```

This investigation targets collector version 0.149.0, tailing Kubernetes CRI container logs in the `/var/log/pods/` directory structure.

Repository: `github.com/open-telemetry/opentelemetry-collector-contrib`
Package: `pkg/stanza`
Feature gate ID: `stanza.synchronousLogEmitter`
Reference issue: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/35456

## Understand the Architecture

The `filelog` receiver uses the stanza framework to tail log files. The data path is:

1. **FileConsumer** in `fileconsumer/` polls the filesystem every `200ms`, configurable with `poll_interval`. For each matched file, a `Reader` goroutine scans new lines since the last poll.

> 2. **Reader batching** - The reader accumulates tokens, that is, scanned lines, into a slice capped at `DefaultMaxBatchSize = 100`. When it reaches EOF or hits `100` tokens, it calls `emitFunc`, a callback wired to `Input.emitBatch`. There is no minimum batch size and no flush timer at this level.

Reviewer 2: ```suggestion
2. **Reader batching** - The reader accumulates tokens (scanned lines) into a slice capped at `DefaultMaxBatchSize = 100`. When it reaches EOF or hits `100` tokens, it calls `emitFunc`, a callback wired to `Input.emitBatch`. There is no minimum batch size and no flush timer at this level.
```

5. **Adapter emitter** - The final operator in the pipeline. The feature gate controls its behavior:
    - **Gate off (default)** - `BatchingLogEmitter` appends entries to an in-memory `[]*entry.Entry` slice and flushes when the buffer reaches `100` entries or every `100ms` using a background goroutine.
    - **Gate on** - `SynchronousLogEmitter` immediately forwards entries to the consumer with no buffering.

6. **receiver.consumeEntries** - Converts `[]*entry.Entry` to `plog.Logs` and calls `ConsumeLogs` on the downstream OTel pipeline consisting of processors and exporters.

### Examine the Container Parser's Internal Sub-Pipeline

When you configure the `container` operator, used for CRI log format parsing, it creates an internal sub-pipeline for handling CRI partial lines with `P` and `F` flags:

```
[recombine operator] → [internal BatchingLogEmitter] → consumeEntries → WriteBatch to main pipeline
```

- The **recombine** operator holds `P` partial lines until a matching `F` final line arrives, then flushes the combined entry.

> - The **internal BatchingLogEmitter** named `criLogEmitter` is hardcoded as `helper.NewBatchingLogEmitter(set, p.consumeEntries)` with default settings of a `100`-entry buffer and `100ms` flush timer. It does not check the `stanza.synchronousLogEmitter` feature gate.

Reviewer 2: active voice

```suggestion
- The container parser hardcodes the internal `BatchingLogEmitter`, named `criLogEmitter`, as `helper.NewBatchingLogEmitter(set, p.consumeEntries)` with a default 100-entry buffer and a 100ms flush timer. It does not check the `stanza.synchronousLogEmitter` feature gate.
```

- The recombine operator also has a background timer called `forceFlushTimeout` that force-flushes incomplete partial lines using `Write` singular to the `criLogEmitter`.

> This internal emitter is the same `BatchingLogEmitter` type used at the adapter level, with the same struct, the same package `operator/helper`, and the same async buffer behavior.

Reviewer 2: active voice

```suggestion
This internal emitter uses the same `BatchingLogEmitter` type as the adapter, with the same struct, package (`operator/helper`), and asynchronous buffer behavior.
```

> This internal emitter is the same `BatchingLogEmitter` type used at the adapter level, with the same struct, the same package `operator/helper`, and the same async buffer behavior.

Reviewer 3: Just curious? If `container` stanza parser can deliver incomplete partial logs, who stitches them together?

Author: No one, so if the partial logs are force flushed, they are just treated as a complete log for the rest of the pipeline

## Describe the Test Methodology

All tests ran with Docker containers on a local machine tailing simulated Pod logs from a log generator using the `filelog` receiver. The backend was a separate OTel Collector instance with an OTLP gRPC receiver.

> The following metrics were observed:

Reviewer 2: ```suggestion
We observed the following metrics:
```

- `otelcol_exporter_sent_log_records_total` - Total records delivered
- `otelcol_exporter_enqueue_failed_log_records_total` - Records dropped due to full exporter queue
- `otelcol_receiver_accepted_log_records_total` - Records accepted by the receiver
- `otelcol_receiver_refused_log_records_total` - Records refused by the receiver, that is, records lost
- `otelcol_process_cpu_seconds_total` and `otelcol_process_runtime_total_alloc_bytes_total` - Resource usage for CPU and memory allocation

The test parameters were:

- Same binary version, 0.149.0 `otelcol-contrib`, for both modes
- Feature gate toggled with `--feature-gates=stanza.synchronousLogEmitter`
- 10 open log files, low-to-moderate write rate
- Log size of ~150 bytes
- Exporter queue size set to `1024`, min and max batch size to `1024`, flush timeout to `10s`, to match batching configurations of other components
- Configurations with full operator chain and with no operators using a raw pipeline

## Present Findings

This section covers the results, test setup, and configuration for each finding.

### Finding 1: With Container Parser, the Feature Gate Has No Observable Effect

Test setup:
- 10 Pods
- 1 log/s per Pod

> Multiple tests on v0.149.0 with the full operator chain, including the container parser, showed identical results between the two emitter modes:

Reviewer 2: ```suggestion
Multiple tests on v0.149.0 with the full operator chain (including the container parser) produced identical results for both emitter modes:
```

### Finding 2: Feature Gate Effect on RPC Calls Depends on Log Rate per Pod

Test configuration: all stanza operators removed, raw fileconsumer to adapter emitter to exporter.

**Low throughput: 1 log/s per Pod, 10 Pods, 300s**

Total lines written: 2930

| Metric                             | Sync  | Batching |
|------------------------------------|-------|----------|
| Export requests (rpc_server count) | 2,930 | 297      |
| Records delivered                  | 2930  | 2930     |
| Avg records / filelog batch        | 1.0   | 9.8      |

**High throughput: 100 logs/s per Pod, 10 Pods, 2146s, queue saturated**

Total lines written: 147956

| Metric                             | Sync   | Batching |
|------------------------------------|--------|----------|
| Export requests (rpc_server count) | 2389   | 2259     |
| Records delivered                  | 147256 | 146656   |
| Avg records / filelog batch        | 61.6   | 65       |

At low throughput, readers rarely fill their 100-entry batch, so the `SynchronousLogEmitter` passes through many small batches while the `BatchingLogEmitter` coalesces them, producing a ~10x difference in export requests.

> At high throughput, readers saturate at ~100 entries per batch regardless of emitter mode, collapsing the difference. The number of records delivered differs due to backpressure and lost data.

Reviewer 2: I assume the data is lost because of backpressure?

```suggestion
At high throughput, readers saturate at ~100 entries per batch regardless of emitter mode, collapsing the difference. The number of records delivered differs due to backpressure causing data loss.
```

### Finding 3: The Container Parser's Internal Emitter Is a Blind Spot for the Feature Gate

> The `stanza.synchronousLogEmitter` feature gate is only checked in one place, `adapter/factory.go` line 79. The container parser hardcodes `helper.NewBatchingLogEmitter` in `operator/parser/container/config.go` line 84 without checking the gate.

Reviewer 2: active voice

```suggestion
The code checks the `stanza.synchronousLogEmitter` feature gate in only one place: `adapter/factory.go` line 79. The container parser hardcodes `helper.NewBatchingLogEmitter` in `operator/parser/container/config.go` line 84 without checking the gate.
```

| **Export requests (rpc_server count)** | **30** | **30**   |
| Avg records / exporter batch           | 97.6   | 97.6     |

The batcher's `flush_timeout` accumulates whatever arrives within `10s` and sends it as one request, coalescing 2,930 individual queue items into 30 gRPC calls, or about 98 log records per batch.

**High throughput: 100 logs/s per Pod, 10 Pods, 300s, queue saturated**

| Metric                                 | Sync    | Batching |
|----------------------------------------|---------|----------|
| Total batches                          | 2389    | 2259     |
| Avg records / filelog batch            | 61.6    | 65       |
| **Export requests (rpc_server count)** | **148** | **148**  |
| Avg records / exporter batch           | 994     | 992      |

At high throughput, readers saturate at ~100 entries per batch regardless of emitter mode, so the difference in batch count collapses and gRPC calls converge naturally.

With the exporter batcher configured, or at high enough throughput, the feature gate is irrelevant to what the backend observes. The only remaining distinctions between emitter modes are the following:

> - **Crash safety** - The synchronous emitter has no in-memory buffer that can lose data. However, the exporter batcher itself has an in-memory buffer with entries waiting for the `200ms` flush, which is also lost on non-graceful shutdown. A persistent queue with `sending_queue.storage: file_storage/...` mitigates this.

Reviewer 2: Gemini says that "the 200ms flush" contradicts the 10s flush_timeout used in this test; suggesting the following to prevent inconsistency.
Also, "which" is ambiguous.

```suggestion
- **Crash safety** - The synchronous emitter has no in-memory buffer that can lose data. However, the exporter batcher itself holds an in-memory buffer of entries waiting for the flush timeout, and this buffer is also lost on non-graceful shutdown. A persistent queue with `sending_queue.storage: file_storage/...` mitigates this.
```

Author: Those are two different flush mechanism, the 10s flush_timeout is the flushing that happens in another component, the exporter batcher. But I guess the details aren't too relevant here, but I'll rephrase it so that this doesn't reference the flush_timeout

| Queue batch send size avg   | 1.0            | 10.0                |
| gRPC calls                  | 30             | 30                  |
| CPU total                   | 7.03s          | 6.90s               |
| `total_alloc`               | 438 MB         | 420 MB              |
| `total_alloc` rate          | ~1.30 MB/s     | ~1.24 MB/s          |

**High throughput: 50 logs/s per Pod, 10 Pods, 1286s, queue saturated**

| Metric                      | Sync (gate on) | Batching (gate off) |
|-----------------------------|----------------|---------------------|
| Queue batch send size count | 1,495          | 1,344               |
| Queue batch send size avg   | 88.5           | 99.4                |
| gRPC calls                  | 96             | 95                  |
| CPU total                   | 2.99s          | 4.82s               |
| `total_alloc`               | 256 MB         | 228 MB              |
| `total_alloc` rate          | ~199 KB/s      | ~177 KB/s           |

At low throughput, both emitter modes are virtually identical in resource consumption. The small differences of 1.3% CPU and 4% allocation are within noise and do not constitute a meaningful finding.

> At high throughput with a saturated queue, CPU usage is 61% higher on the batching emitter, 2.99s compared to 4.82s. This is consistent across two independent runs and is attributable to the batching emitter's flusher goroutine repeatedly waking up every `100ms`, acquiring the buffer lock, and attempting `Offer` against a full queue, doing real work that produces nothing. The sync emitter has no such goroutine. Backpressure is absorbed by the reader goroutines blocking, which costs no CPU.

Reviewer 2: trying to split the long-running sentence

```suggestion
At high throughput with a saturated queue, CPU usage is 61% higher on the batching emitter (2.99s compared to 4.82s). This behavior is consistent across two independent runs. It occurs because the batching emitter's flusher goroutine repeatedly wakes up every 100ms, acquires the buffer lock, and attempts to call `Offer` against a full queue, which performs work but produces no result. The sync emitter has no such goroutine. Backpressure is absorbed by the reader goroutines blocking, which costs no CPU.
```

The `total_alloc` rate is slightly higher on sync under saturation, 199 compared to 177 KB/s, but the difference is small and negligible.

The dominant resource conclusion is that the batching emitter wastes CPU under queue saturation. Otherwise, the two modes are equivalent.

## Examine Technical Details

This section provides additional technical context on the components discussed in the findings.

### Understand BatchingLogEmitter Flush Behavior

The `BatchingLogEmitter` has two flush triggers:

- **Size** - When `ProcessBatch` or `Process` appends entries and the buffer reaches `maxBatchSize` with a default of `100`, it flushes immediately on the calling goroutine.
- **Time** - A background goroutine ticks every `flushInterval` with a default of `100ms` and flushes whatever has accumulated, even if it is just 1 entry.

> On graceful shutdown with `Stop`, the flusher does a final flush with a 5-second timeout context. On non-graceful shutdown such as SIGKILL or OOM, the buffer is lost.

Reviewer 2: ```suggestion
On graceful shutdown (`Stop`), the flusher performs a final flush with a 5-second timeout context. On non-graceful shutdown (such as SIGKILL or OOM), the buffer is lost.
```

With the `SynchronousLogEmitter`, each reader's batch immediately becomes a separate `ConsumeLogs` call, producing as many export requests as there are active files.

### Compare Pipeline Flow Before and After Feature Gate

<p>
<img src="assets/pipeline_batching_emitter.svg" alt="pipeline-flow-batch" width="600" />
<em>Pipeline Flow with BatchingLogEmitter</em>
</p>

<p>
<img src="assets/pipeline_sync_emitter.svg" alt="pipeline-flow-sync" width="600" />
<em>Pipeline Flow with SynchronousLogEmitter</em>
</p>

Without the feature gate, on the left, the container parser's internal `BatchingLogEmitter` runs and coalesces entries tailed from all readers into one batch. After that, the adapter's `BatchingLogEmitter` buffers it again up to `100` entries and then flushes them together.

With the feature gate enabled, on the right, the container parser still runs `BatchingLogEmitter` internally, but the adapter delivers entries synchronously to the next consumer instead of batching them into a buffer.

> The `BatchingLogEmitter` used internally by the container parser behaves exactly the same as the one in the adapter. In the left diagram, the second `BatchingLogEmitter` therefore has little additional effect.

Reviewer 2: ```suggestion
The `BatchingLogEmitter` that the container parser uses internally behaves exactly like the one in the adapter. In the left diagram, the second `BatchingLogEmitter` therefore has little additional effect.
```

## Conclusions

This section summarizes the investigation results and provides recommendations.

### Use the Filelog Receiver with the Container Parser

The feature gate is safe to enable. It provides theoretical crash-safety and backpressure benefits with no observable impact on throughput, export request count, or resource usage, because the container parser's internal emitter already handles the coalescing.

Enabling the synchronous emitter significantly increases the number of export requests to the backend, depending on the number of files tailed and the log production rate. If the backend rate-limits by request count, this matters. Enabling batching in the exporter mitigates this.

### Enable the Exporter Batcher with Synchronous Emitter

The exporter batcher `sending_queue.batch` makes the feature gate choice irrelevant to backend behavior. With the batcher enabled, both emitter modes produce the same number of gRPC requests and the same batch sizes at the wire level. The feature gate remains useful for crash safety and backpressure semantics, but has no impact on export request count or backend load.

> Additionally, enabling the `SynchronousLogEmitter` with an exporter batcher reduces CPU usage under high load scenarios as shown in [Finding 5](#finding-5-the-collector-resource-consumption-changes-with-the-feature-gate-under-heavy-load).

Reviewer 2: ```suggestion
Additionally, combining the `SynchronousLogEmitter` with an exporter batcher reduces CPU usage under high load (see [Finding 5](#finding-5-the-collector-resource-consumption-changes-with-the-feature-gate-under-heavy-load)).
```
